hathor/cli/side_dag.py

```diff
@@ -0,0 +1,220 @@
+        return
+
+    assert isinstance(message, HathorProcessInitSuccess)
+    logger.info('hathor node initialized')
+    logger.info('starting side-dag node...')
+
+    try:
+        side_dag = SideDagRunNode(argv=argv)
+    except (BaseException, Exception):
+        logger.critical('terminating hathor node...')
+        conn.send(SideDagProcessTerminated())
+        hathor_node_process.terminate()
+        return
+
+    side_dag.run()
+
+    # If `run()` returns, either the hathor node exited and terminated us, leaving the message below, or the side-dag
+    # node exited and we will terminate the hathor node.
+    if conn.poll():
+        message = conn.recv()
+        assert isinstance(message, HathorProcessTerminated)
+        logger.critical('side-dag node terminated because hathor node exited')
+        return
+
+    conn.send(SideDagProcessTerminated())
+    logger.critical('terminating hathor node...')
+    hathor_node_process.terminate()
+
+
+def _start_hathor_node_process(argv: list[str], *, capture_stdout: bool, conn: 'Connection') -> Process:
+    """Create and start a Hathor node process."""
+    run_hathor_node_args = (argv, RunNode, capture_stdout, conn)
+    hathor_node_process = Process(target=_run_hathor_node, args=run_hathor_node_args)
+    hathor_node_process.start()
+    return hathor_node_process
+
+
+def _run_hathor_node(argv: list[str], run_node_cmd: type[RunNode], capture_stdout: bool, conn: 'Connection') -> None:
+    """Function to be called by a background process to run the Hathor full node."""
+    from hathor.cli.util import process_logging_options, setup_logging
+
+    # We don't terminate via SIGINT directly, instead the side-dag process will terminate us.
+    signal.signal(signal.SIGINT, lambda _, __: None)
+    try:
+        log_options = process_logging_options(argv)
+        setup_logging(
+            debug=log_options.debug,
+            capture_stdout=capture_stdout,
+            logging_output=log_options.logging_output,
+            sentry=log_options.sentry,
+        )
+        hathor_node = run_node_cmd(argv=argv)
+    except (BaseException, Exception):
+        conn.send(HathorProcessInitFail(traceback.format_exc()))
+        return
+
+    conn.send(HathorProcessInitSuccess())
+    hathor_node.run()
+
+    # If `run()` returns, either the side-dag node exited and terminated us, leaving the message below, or the hathor
+    # node exited and we will terminate the side-dag node.
+    if conn.poll():
+        message = conn.recv()
+        assert isinstance(message, SideDagProcessTerminated)
+        logger.critical('hathor node terminated because side-dag process was terminated')
+        return
+
+    conn.send(HathorProcessTerminated())
+    logger.critical('terminating side-dag node...')
+    parent_pid = os.getppid()
+    parent_process = psutil.Process(parent_pid)
+    parent_process.terminate()
```

Comment:
What does `terminate` do? Does it just send a `SIGTERM`?

Reply:
> What does `terminate` do? Does it just send a `SIGTERM`? Because that alone might not be enough to reliably terminate the parent process

Yes, it just sends a `SIGTERM`. Why not?

> there's an issue about doing a monitor process in a separate PR, that's fine by me.

Yep, it will be a future improvement.

Reply:
I added this in the issue: https://github.com/HathorNetwork/hathor-core/issues/1051#issuecomment-2153518597

Reply:
@jansegre I had a similar conversation about this topic with @glevco during my review.